x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java

```diff
@@ -53,11 +76,85 @@ public class IlmHealthIndicatorService implements HealthIndicatorService {
+                + "could be impacted.",
+            List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)
+        )
+    );
+
+    public static final List<HealthIndicatorImpact> STAGNATING_INDEX_IMPACT = List.of(
+        new HealthIndicatorImpact(
+            NAME,
+            STAGNATING_INDEX_IMPACT_ID,
+            3,
+            "Automatic index lifecycle and data retention management cannot make progress on one or more indices. The performance and "
+                + "stability of the indices and/or the cluster could be impacted.",
+            List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)
+        )
+    );
+
+    private static final TimeValue ONE_DAY = TimeValue.timeValueDays(1);
+
+    static final Map<String, RuleConfig> RULES_BY_ACTION_CONFIG = Map.of(
+        RolloverAction.NAME,
+        actionRule(RolloverAction.NAME).stepRules(stepRule(WaitForActiveShardsStep.NAME, ONE_DAY)),
+        //
+        MigrateAction.NAME,
+        actionRule(MigrateAction.NAME).maxTimeOnAction(ONE_DAY).noStepRules(),
+        //
+        SearchableSnapshotAction.NAME,
+        actionRule(SearchableSnapshotAction.NAME).maxTimeOnAction(ONE_DAY)
+            .stepRules(
+                stepRule(WaitForDataTierStep.NAME, ONE_DAY),
+                stepRule(WaitForIndexColorStep.NAME, ONE_DAY),
+                stepRule(WaitForNoFollowersStep.NAME, ONE_DAY)
+            ),
+        //
+        DeleteAction.NAME,
+        actionRule(DeleteAction.NAME).maxTimeOnAction(ONE_DAY).stepRules(stepRule(DeleteStep.NAME, ONE_DAY)),
+        //
+        ShrinkAction.NAME,
+        actionRule(ShrinkAction.NAME).maxTimeOnAction(ONE_DAY).stepRules(stepRule(WaitForNoFollowersStep.NAME, ONE_DAY)),
+        //
+        AllocateAction.NAME,
+        actionRule(AllocateAction.NAME).maxTimeOnAction(ONE_DAY).noStepRules(),
+        //
+        ForceMergeAction.NAME,
+        actionRule(ForceMergeAction.NAME).maxTimeOnAction(ONE_DAY).stepRules(stepRule(WaitForIndexColorStep.NAME, ONE_DAY)),
+        //
+        DownsampleAction.NAME,
+        actionRule(DownsampleAction.NAME).maxTimeOnAction(ONE_DAY).stepRules(stepRule(WaitForNoFollowersStep.NAME, ONE_DAY))
+    );
+
+    public static final Collection<RuleConfig> ILM_RULE_EVALUATOR = RULES_BY_ACTION_CONFIG.values();
+
+    static final Map<String, Diagnosis.Definition> STAGNATING_ACTION_DEFINITIONS = RULES_BY_ACTION_CONFIG.entrySet()
+        .stream()
+        .collect(
+            Collectors.toUnmodifiableMap(
+                Map.Entry::getKey,
+                entry -> new Diagnosis.Definition(
+                    NAME,
+                    "stagnating_action:" + entry.getKey(),
+                    "Some indices have been stagnated on the action [" + entry.getKey() + "] longer than the expected time.",
+                    "Check the current status of the Index Lifecycle Management service using the [/_ilm/explain] API.",
```

Comment:
```suggestion
                    "Check the current status of the Index Lifecycle Management for every affected index using the [GET /_ilm/explain] API.",
```
I wonder if we should make it explicit that the API is called on the index itself

Maybe `[GET /affected_index_name/_ilm/explain] API. Please replace the affected_index_name in the API with the actual index name.`

What do you think?